### PR TITLE
ordering output transfers

### PR DESCRIPTION
**Pull request by matei-p**

### CODE REVIEW

**Review by sasurobert — DISMISSED** [review]

File: output.go [hunk 1]
```diff
@@ -184,6 +186,29 @@ func (vmOutput *VMOutput) GetFirstReturnData(asType vm.ReturnDataKind) (interfac
...
+				maxTransferIndex = int(transfer.Index)
+			}
+		}
+	}
+	return uint32(maxTransferIndex + 1)
+}
+
+// ReindexTransfers from VMOutput
+func (vmOutput *VMOutput) ReindexTransfers(nextIndexProvider NextOutputTransferIndexProvider) {
+	for _, account := range vmOutput.OutputAccounts {
```
Comment (sasurobert) [inline output.go:204]: this is not deterministic. as the iteration on accounts is done in a map. Why is this function needed ?

**Review by sasurobert — CHANGES_REQUESTED** [review]

File: interface.go [hunk 2]
```diff
@@ -405,3 +405,7 @@ type GuardedAccountHandler interface {
 	CleanOtherThanActive(uah UserAccountHandler)
 	IsInterfaceNil() bool
 }
+
+type NextOutputTransferIndexProvider interface {
+	NextOutputTransferIndex() uint32
```
Comment (sasurobert) [inline interface.go:410]: add IsInterfaceNil

File: output.go [hunk 3]
```diff
@@ -184,6 +186,29 @@ func (vmOutput *VMOutput) GetFirstReturnData(asType vm.ReturnDataKind) (interfac
...
+			if int(transfer.Index) > maxTransferIndex {
+				maxTransferIndex = int(transfer.Index)
+			}
+		}
+	}
+	return uint32(maxTransferIndex + 1)
+}
+
+// ReindexTransfers from VMOutput
+func (vmOutput *VMOutput) ReindexTransfers(nextIndexProvider NextOutputTransferIndexProvider) {
```
Comment (sasurobert) [inline output.go:203]: add checkIfNil for nextIndexProvider

File: output.go [hunk 4]
```diff
@@ -184,6 +186,29 @@ func (vmOutput *VMOutput) GetFirstReturnData(asType vm.ReturnDataKind) (interfac
 	return nil, fmt.Errorf("can't interpret return data")
 }
 
+// GetMaxOutputTransferIndex returns the maximum output transfer index
+func (vmOutput *VMOutput) GetNextAvailableOutputTransferIndex() uint32 {
+	maxTransferIndex := -1
+	for _, account := range vmOutput.OutputAccounts {
+		for _, transfer := range account.OutputTransfers {
+			if int(transfer.Index) > maxTransferIndex {
```
Comment (sasurobert) [inline output.go:194]: why do not keep everything as uint32 ?

**Review by sasurobert — CHANGES_REQUESTED** [review]

File: interface.go [hunk 5]
```diff
@@ -405,3 +405,9 @@ type GuardedAccountHandler interface {
 	CleanOtherThanActive(uah UserAccountHandler)
 	IsInterfaceNil() bool
 }
+
+type NextOutputTransferIndexProvider interface {
```
Comment (sasurobert) [inline interface.go:409]: add comment for public interface

File: output.go [hunk 6]
```diff
@@ -184,6 +186,57 @@ func (vmOutput *VMOutput) GetFirstReturnData(asType vm.ReturnDataKind) (interfac
...
+		return 0
+	}
+
+	return maxTransferIndex + 1
+}
+
+// ReindexTransfers from VMOutput
+func (vmOutput *VMOutput) ReindexTransfers(nextIndexProvider NextOutputTransferIndexProvider) error {
+
+	if nextIndexProvider == nil {
```
Comment (sasurobert) [inline output.go:212]: if check.IfNil(nextIndexProvider)

File: output.go [hunk 7]
```diff
@@ -184,6 +186,57 @@ func (vmOutput *VMOutput) GetFirstReturnData(asType vm.ReturnDataKind) (interfac
...
+	firstAccountWithTransfers := true
+	for _, account := range vmOutput.OutputAccounts {
+		if len(account.OutputTransfers) != 0 {
+			if firstAccountWithTransfers {
+				firstAccountWithTransfers = false
+				continue
+			}
+
+			// this is not the first account with transfers and the others were not indexed
+			if nextIndexProvider.GetCrtTransferIndex() == 0 {
```
Comment (sasurobert) [inline output.go:225]: you should check first whether the outPutTransfers have been indexed. If yes, then it is all good.

**Review by andrei-marinica — APPROVED** [review]

**Pull request merged**